Mathlib/Data/Finset/Density.lean

```diff
@@ -0,0 +1,217 @@
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Order.Field.Basic
+import Mathlib.Tactic.Positivity.Basic
+import Mathlib.Tactic.Positivity.Finset
+
+/-!
+# Density of a finite set
+
+This defines the density of a `Finset` and provides induction principles for finsets.
+
+## Main declarations
+
+* `Finset.dens t`: `dens s : ℕ` returns the density of `s : Finset α`.
+
+## Notation
+
+If you need to specify the field the density is valued in, `dens[𝕜] s` is notation for
+`(dens s : 𝕜)`. Note that no dot notation is allowed.
+-/
+
+open Function Multiset Nat
+
+variable {𝕜 α β : Type*} [Fintype α]
+
+namespace Finset
+section Semifield
+variable [Semifield 𝕜] {s t : Finset α} {a b : α}
+
+/-- `dens s` is the number of elements of `s`, aka its density. -/
+@[pp_dot] def dens (s : Finset α) : 𝕜 := s.card / Fintype.card α
+
+@[inherit_doc dens] notation "dens[" 𝕜 "]" => @dens 𝕜
+
+lemma card_div_card_eq_dens (s : Finset α) : dens[𝕜] s = s.card / Fintype.card α := rfl
+
+@[simp] lemma dens_empty : dens[𝕜] (∅ : Finset α) = 0 := by simp [dens]
+
+@[simp] lemma dens_singleton (a : α) : dens ({a} : Finset α) = (Fintype.card α : 𝕜)⁻¹ := by
+  simp [dens]
+
+@[simp]
+lemma dens_cons (h : a ∉ s) : (s.cons a h).dens = dens s + (Fintype.card α : 𝕜)⁻¹ := by
+  simp [dens, add_div]
+
+@[simp]
+lemma dens_disjUnion (s t : Finset α) (h) : dens[𝕜] (s.disjUnion t h) = dens s + dens t := by
+  simp_rw [dens, card_disjUnion, Nat.cast_add, add_div]
+
+section Lattice
+variable [DecidableEq α]
+
+lemma dens_union_add_dens_inter (s t : Finset α) :
+    dens[𝕜] (s ∪ t) + dens (s ∩ t) = dens s + dens t := by
+  simp_rw [dens, ← add_div, ← Nat.cast_add, card_union_add_card_inter]
+
+lemma dens_inter_add_dens_union (s t : Finset α) :
+    dens[𝕜] (s ∩ t) + dens (s ∪ t) = dens s + dens t := by rw [add_comm, dens_union_add_dens_inter]
+
+@[simp] lemma dens_union_of_disjoint (h : Disjoint s t) : dens[𝕜] (s ∪ t) = dens s + dens t := by
+  rw [← disjUnion_eq_union s t h, dens_disjUnion _ _ _]
+
+lemma dens_sdiff_add_dens_eq_dens (h : s ⊆ t) :  dens[𝕜] (t \ s) + dens s = dens t := by
+  simp [dens, ← card_sdiff_add_card_eq_card h, add_div]
+
+lemma dens_sdiff_add_dens : dens[𝕜] (s \ t) + dens t = (s ∪ t).dens := by
+  rw [← dens_union_of_disjoint sdiff_disjoint, sdiff_union_self_eq_union]
+
+lemma dens_sdiff_comm [IsCancelAdd 𝕜] (h : card s = card t) : dens[𝕜] (s \ t) = dens (t \ s) :=
+  add_left_injective (dens t) $ by
+    simp_rw [dens_sdiff_add_dens, union_comm s, ← dens_sdiff_add_dens, dens, h]
+
+@[simp]
+lemma dens_sdiff_add_dens_inter (s t : Finset α) : dens[𝕜] (s \ t) + dens (s ∩ t) = dens s := by
+  rw [← dens_union_of_disjoint (disjoint_sdiff_inter _ _), sdiff_union_inter]
+
+@[simp]
+lemma dens_inter_add_dens_sdiff (s t : Finset α) : dens[𝕜] (s ∩ t) + dens (s \ t) = dens s := by
+  rw [add_comm, dens_sdiff_add_dens_inter]
+
+lemma dens_filter_add_dens_filter_not_eq_dens
+    (p : α → Prop) [DecidablePred p] [∀ x, Decidable (¬p x)] :
```

Comment:
```suggestion
    (p : α → Prop) [DecidablePred p] :
```

Reply:
That's less syntactically general.

Reply:
Good point

Reply:
I thought about this a little more. Doesn't removing an extra unnecessary assumption make your theorem more general, not less general?

Reply:
No, not here. Inspect your version vs my version in the infoview to make up your mind.

Hint: Removing the assumption is basically equivalent to saying "`∀ a, a + a = a + a` is more general than `∀ a b, a = b → a + b = b + a` because it doesn't have the `b` assumption". This is true mathematically (although note that saying "`b` assumption" is mathematically dodgy), but in practice it makes your theorem less applicable because when you want to rewrite some `a + b` found in nature you need Lean to be able to *unify* `a` and `b` syntactically, instead of being able to provide the proof yourself.

Reply:
I see.